src/main/java/com/google/api/codegen/transformer/DynamicLangGapicSamplesTransformer.java

```diff
@@ -0,0 +1,258 @@
+            .collect(ImmutableList.toImmutableList());
+
+    ImmutableTable<String, String, ImmutableList<SampleConfig>> sampleConfigTable =
+        productConfig.getSampleConfigTable();
+
+    // We don't have sample configs written in sample config. Continue to use gapic config.
+    if (sampleConfigTable.isEmpty()) {
+      return generateSamplesFromGapicConfigs(interfaceContexts, productConfig, namer);
+    }
+
+    // Generate samples using sample configs.
+    return generateSamplesFromSampleConfigs(
+        interfaceContexts, productConfig, namer, sampleConfigTable);
+  }
+
+  @Override
+  public List<String> getTemplateFileNames() {
+    return Arrays.asList(templateFileName);
+  }
+
+  private List<ViewModel> generateSamplesFromGapicConfigs(
+      List<InterfaceContext> interfaceContexts,
+      GapicProductConfig productConfig,
+      SurfaceNamer namer) {
+    List<MethodSampleView> allSamples =
+        interfaceContexts
+            .stream()
+            .flatMap(c -> apiMethodTransformer.generateApiMethods(c).stream())
+            .flatMap(m -> m.samples().stream())
+            .collect(ImmutableList.toImmutableList());
+    SampleFileRegistry registry = new SampleFileRegistry(namer, allSamples);
+    ImmutableList.Builder<ViewModel> sampleFileViews = ImmutableList.builder();
+    for (InterfaceContext context : interfaceContexts) {
+      List<OptionalArrayMethodView> methods = apiMethodTransformer.generateApiMethods(context);
+      for (OptionalArrayMethodView method : methods) {
+        for (MethodSampleView sample : method.samples()) {
+          sampleFileViews.add(
+              newSampleFileView(
+                  productConfig,
+                  context,
+                  registry.getSampleFileName(sample, method.name()),
+                  method,
+                  sample));
+        }
+      }
+    }
+    return sampleFileViews.build();
+  }
+
+  private List<ViewModel> generateSamplesFromSampleConfigs(
+      List<InterfaceContext> interfaceContexts,
+      GapicProductConfig productConfig,
+      SurfaceNamer namer,
+      ImmutableTable<String, String, ImmutableList<SampleConfig>> sampleConfigTable) {
+
+    Map<SampleConfig, Integer> configNamesCount = new HashMap<>();
+    for (InterfaceContext interfaceContext : interfaceContexts) {
+      for (MethodModel method : interfaceContext.getSupportedMethods()) {
+        MethodContext methodContext = interfaceContext.asRequestMethodContext(method);
+        String interfaceName =
+            interfaceContext.getInterfaceConfig().getInterfaceModel().getFullName();
+        String methodName = method.getSimpleName();
+        MoreObjects.firstNonNull(
+                sampleConfigTable.get(interfaceName, methodName), ImmutableList.<SampleConfig>of())
+            .stream()
+            .forEach(configNamesCount.put(c, namer.getMatchingCallingForms(methodContext, c.callingPattern())))
```

Comment:
I'm confused by this sequence. Could you add some clarifying comments? In particular, you're getting a list of matching calling forms, and then you're calling a very similar but separate function to see whether there are any matching calling forms, and then you're putting something again? 

A loop might be simpler
(see, eg, https://blog.jooq.org/2015/12/08/3-reasons-why-you-shouldnt-replace-your-for-loops-by-stream-foreach/)

Reply:
Sorry I run into some problem when add PHP test, but I had to leave yesterday so pushed some WIP code here so that I can catch up today WFH. I'll let you know when it's ready.